objc_msgSend.aarch64.S

```diff
@@ -1,12 +1,78 @@
 #define ARGUMENT_SPILL_SIZE (8*10 + 8*16)
-.macro MSGSEND receiver, sel
-	.cfi_startproc
+
+/* Windows ARM64 Exception Handling
+ *
+ * Structured Exception Handling (SEH) on Windows ARM64 differs from the x64
+ * implementation. Functions consist of a single prologue and zero or more
+ * epilogues. Instead of using offsets for the .seh* directives to manipulate the
+ * stack frame, each directive corresponds to a single instruction.
+ *
+ * This presents a challenge for our objc_msgSend function, which only modifies
+ * the stack when a slow lookup is needed (see label "5").
+ *
+ * To address this, we move the directive marking the start of a function deep
+ * into the msgSend body to prevent marking every instruction as ".seh_nop."
+ *
+ *   For 64-bit Windows (_WIN64):
+ *  - EH_START(x): Start of function (no effect on Windows)
+ *  - EH_END(x): End of function (no effect on Windows)
+ *  - EH_START_AT_OFFSET(x): Mark Start of function (Delayed)
+ *  - EH_END_AT_OFFSET(x): Mark End of function (Delayed)
+ *  - EH_END_PROLOGUE: End of function prologue
+ *  - EH_START_EPILOGUE: Start of function epilogue
+ *  - EH_END_EPILOGUE: End of function epilogue
+ *  - EH_SAVE_FP_LR(x): Save Frame Pointer and Link Register
+ *  - EH_STACK_ALLOC(x): Stack allocation (inside prologue)
+ *  - EH_ADD_FP(x): Add to Frame Pointer
+ *  - EH_NOP: Mark instruction with no unwinding relevance
+ *
+ * For non-64-bit Windows systems or other platforms, these macros have no effect and can be used without causing issues.
+ */
+
+#ifdef _WIN64
+#   define EH_START(x)
+#   define EH_END(x)
+
+#   define EH_START_AT_OFFSET(x) .seh_proc x
+#   define EH_END_AT_OFFSET(x) .seh_endproc x
+
+#   define EH_END_PROLOGUE .seh_endprologue
+#   define EH_START_EPILOGUE .seh_startepilogue
+#   define EH_END_EPILOGUE .seh_endepilogue
+
+#   define EH_SAVE_FP_LR(x) .seh_save_fplr x
+#   define EH_STACK_ALLOC(x) .seh_stackalloc x
+#   define EH_ADD_FP(x) .seh_add_fp	x
+
+#   define EH_NOP .seh_nop
+#else
+// Marks the real start and end of the function
+#   define EH_START(x) .cfi_startproc
+#   define EH_END(x) .cfi_endproc
```

Comment:
Neither of the definitions of these seem to use the argument.  Why is it there?

Reply:
You are correct. A similar macro is in the x86_64 implementation (END_PROC) which is not using the arg either. I will fix that!